src/server/ButtplugServer.ts

```diff
@@ -89,7 +95,20 @@ export class ButtplugServer extends EventEmitter {
     this.OnOutgoingMessage(new Messages.Log(LogLevel[aMsg.LogLevel], aMsg.Message));
   }
 
-  private OnOutgoingMessage = (aMsg: Messages.ButtplugMessage) => {
-    this.emit("message", aMsg);
+  private OnOutgoingMessage = (msg: Messages.ButtplugMessage) => {
+    if (msg.constructor.name === "Error") {
+      return msg;
+    }
+    if (this._clientSchemaVersion === -1) {
+      return new Messages.Error("Cannot discern client schema version. Was RequestServerInfo message sent?");
+    }
+    while (msg.SchemaVersion !== this._clientSchemaVersion && msg.SchemaVersion > 0) {
+      msg = msg.DowngradeMessage();
+    }
+    // If there was a conversion problem, log as well as returning an error message.
+    if (msg.constructor.name === "Error") {
+      this._logger.Error((msg as Messages.Error).ErrorMessage);
```

Comment:
What if it wasn't a downgrade error? Devices can reply with errors too.

Reply:
There's an early exit on the returned message already being an Error a few lines up. If it wasn't a downgrade Error, we'll already have bailed.